src/tests/ftest/launch.py

```diff
@@ -20,164 +20,471 @@
+                display("Error reading {}: {}".format(yaml_file, error))
+                exit(1)
+    return yaml_data
+
+
+def clean_test_environment(machines, debug):
+    """Clean the daos test environment on each specified machine.
+
+    Args:
+        machines (list): a list of machines to clean
+        debug (bool): whether or not to dispaly debug messages
+
+    Returns:
+        None
+
     """
-    Launch a given test.
+    # List of processes to kill on each daos server/client if active
+    daos_process_names = [
+        "orterun",
+        "orted",
+        "daos_server",
+        "daos_io_server",
+        "aexpect-helper-2.7",
+        "avocado",
+        "aexpect-helper-",
+        "daos_agent",
+    ]
+
+    # List of files to remove on each daos server/client
+    daos_files = {
+        "server_log": "/tmp/server.log",
+        "agent_log": "/tmp/daos_agent.log",
+        "control_log": "/tmp/daos_control.log",
+        "sockets_dir": "/tmp/daos_sockets",
+    }
+
+    # Replace any default log files/dirs with yaml definitions
+    yaml_data = get_yaml_data("./data/daos_server_baseline.yaml")
+    display("DAOS server yaml data: {}".format(yaml_data), debug)
+    if isinstance(yaml_data, dict):
+        for key, value in yaml_data.items():
+            if isinstance(value, dict):
+                for key2, value2 in yaml_data[key].items():
+                    if key2 == "log_file":
+                        daos_files["server_log"] = value2
+            elif key == "control_log_file":
+                daos_files["control_log"] = value
+            elif key == "sockets_dir":
+                daos_files["sockets_dir"] = value
+
+    clean_commands = [
+        "killall -9 {}".format(" ".join(daos_process_names)),
+        "rm -rf /mnt/daos/*",
+    ]
+    clean_commands.extend(
+        ["rm -rf {}".format(log) for log in daos_files.values()])
+    for machine in machines:
+        if len(machine) > 0:
+            display("Cleaning test environment on {}".format(machine))
+            output = get_output("ssh {} \"{}\"".format(
+                machine, "; ".join(clean_commands)))
+            display("  Output: {}".format(output), debug)
+
+
+def get_test_list(tags):
+    """Generate a list of tests and avocado tag filter from a list of tags.
+
+    Args:
+        tags (list): a list of tag or test file names
+
+    Returns:
+        (str, list): a tuple of the avacado tag filter and lists of tests
+
     """
-    _file = os.path.relpath(_file)
-    param_file = yamlforpy(_file)
-    params = ' --mux-yaml ' + param_file
-    test_cmd = avocado + ignore_errors + output_options
-    if use_tags:
-        test_cmd += category
-    test_cmd += params + ' -- ' + _file
+    tag_filter = ""
+    test_list = []
+    for tag in tags:
+        if ".py" in tag:
+            # Assume '.py' indicates a test and just add it to the list
+            test_list.append(tag)
+        elif tag == "all":
```

Comment:
Good question.  I think one can make the argument that if `all` exists with other tags then it really doesn't matter that the other tags were specified and we understand the user's intent fairly well.  I can also see an argument where this combination doesn't make sense and it's best to ask the user to review the inputs and try again - especially if the intent may have been "run all the tests and then repeat this one or two".  There is also an assumption here that any `*.py` `tags` argument would also include a test that would be included by `all`, which is probably fair for our intended use, but nothing prevents the user from specifying a non ftest test via this method.

My preference would be to remove the special handling of `all` (treating it as a regular tag) and treat the lack of a `tags` argument as meaning do not include the `--filter-by-tags=<tag>` option with the `avocado` command, thus resulting in all tests.

Reply:
So a commit message pragma of:

```
Test-tags:
```

indicates the "all" (by way of not having any `--filter-by-tags=` option?

Reply:
Ideally it would, but it looks like the current code in `Jenkinsfile `would prevent a developer from passing an empty test_tag into ftest.sh and passing quoted empty string into launch.py does not have the same behavior as passing in no tag:

```
$ ./launch.py --list ''
Arguments: Namespace(archive=False, clean=False, list=True, rename=False, sparse=False, tags=[''])
Running avocado list --paginator off --filter-by-tags= ./ | sed -ne '/INSTRUMENTED/s/.* \([^:]*\):.*/\1/p' | uniq
ERROR: No tests or tags found via ['']
```

Maybe we could process `Test-tag: -` as meaning no tags / all tests with a simple Jenkinsfile change:

```
test_tag=$(git show -s --format=%B | sed -ne "/^Test-tag:/s/^.*: *//p")
if [ -z "$test_tag" ]; then
    test_tag=regression,vm
elif [ "$test_tag" == "-" ]; then
    test_tag=""
fi
```

Reply:
Or fix the Jenkinsfile handling to detect a(n i.e. empty) `Test-tag:` pragma by first checking for the pragma and then extracting it's value rather than just blindly trying to extract a value:

```bash
$ read pragma value < <(echo "Test-tag: foo,bar bat" | grep Test-tag:); echo "$pragma->$value"
Test-tag:->foo,bar bat
$ read pragma value < <(echo "Test-tag: foo,bar" | grep Test-tag:); echo "$pragma->$value"
Test-tag:->foo,bar
$ read pragma value < <(echo "Test-tag:      " | grep Test-tag:); echo "$pragma->$value"
Test-tag:->
$ read pragma value < <(echo "Test-tag:" | grep Test-tag:); echo "$pragma->$value"
Test-tag:->
$ read pragma value < <(echo "boo" | grep Test-tag:); echo "$pragma->$value"
->
```

Where obviously you'd instead put the `git show ... | grep` inside the `<()` and then based on what's in `$pragma` pass `$value` as the test tag or an empty tag.

This all said, @skirvan44 has a [PR](https://github.com/daos-stack/daos/pull/574) in progress to do some retagging and is adding an "all" tag to (just about) every test.

Reply:
Upon reflection, I am not sure that an empty Test-tag: representing _all_ is that great an idea.  If you know avocado it makes sense, but I don't think that will apply to the majority of people and they will simply just not realise to get all tests, you leave the tag empty.

So I think I am changing my position and that to run all tests, you specify an **all** tag.  Whether that is handled as a special case as it is currently or by way of adding an _all_ tag to every test, I don't think I really care.